chain/lift.go

```diff
@@ -4,42 +4,55 @@ import (
 	"log"
 )
 
-//TPosToQPos converts a target position in a chain to the corresponding query position.
-func TPosToQPos(c Chain, TPos int) int {
+// TPosToQPos converts a target position in a chain to the corresponding query position.
+// The bool returns true if the returned TPos was found in a (Size) section of a Chain, and
+// false otherwise, as in when the TPos fell in TBases.
+func TPosToQPos(c Chain, TPos int) (int, bool) {
 	var currT, currQ int
 
 	//if target is negative strand, we convert TStart to the reverse complement position.
 	if c.TStrand {
 		currT = c.TStart
 	} else {
-		currT = c.TSize - 1 - c.TStart
+		log.Fatalf("All target strands should be positive. Current chain:\n %s\n", ToString(c))
 	}
 
 	if c.QStrand {
 		currQ = c.QStart
 	} else {
-		currQ = c.QSize - 1 - c.QStart
+		currQ = c.QEnd - 1
 	}
 
 	if TPos < c.TStart || TPos > c.TEnd {
 		log.Fatalf("Error in TPosToQPos: TPos: %d, is not within the range of the chain. TStart: %d. TEnd: %d.", TPos, c.TStart, c.TEnd)
 	}
-
 	for i := 0; i < len(c.Alignment); i++ {
-		if currT+c.Alignment[i].Size > TPos {
-			return currQ + TPos - currT
-		}
-		currT += c.Alignment[i].Size
-		currQ += c.Alignment[i].Size
-		if currT+c.Alignment[i].TBases > TPos {
-			//in this case, the TPos is in a place with no corresponding query block.
-			//Should we just return currQ?
-			return currQ
+		if c.QStrand {
+			if currT+c.Alignment[i].Size > TPos {
+				return currQ + (TPos - currT), true
+			}
+			currT += c.Alignment[i].Size
+			currQ += c.Alignment[i].Size
+			if currT+c.Alignment[i].TBases > TPos {
+				//in this case, the TPos is in a place with no corresponding query block.
+				return currQ, false
+			}
+			currT += c.Alignment[i].TBases
+			currQ += c.Alignment[i].QBases
+		} else {
+			if currT+c.Alignment[i].Size > TPos {
+				return currQ - (TPos - currT), true
+			}
+			currT += c.Alignment[i].Size
+			currQ -= c.Alignment[i].Size
+			if currT+c.Alignment[i].Size > TPos {
```

Comment:
Should this be:
currT+c.Alignment[i].TBases
instead of:
currT+c.Alignment[i].Size
?